### PR TITLE
Improvements to the background compiler component

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Improvements documented inline.

### CODE REVIEW

File: src/Features/Core/Portable/Workspace/BackgroundCompiler.cs [hunk 1]
```diff
@@ -8,30 +8,36 @@
...
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Host
 {
     internal sealed class BackgroundCompiler : IDisposable
     {
-        private Workspace _workspace;
-        private readonly TaskQueue _taskQueue;
+        private Workspace? _workspace;
+        private readonly AsyncBatchingWorkQueue<CancellationToken> _workQueue;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Workspace/BackgroundCompiler.cs:20]: first: using a AsyncBatching queue to debounce things.  We use near-immediate so we can start running extremely quickly.  But this still allows us to batch up massive flurries of events without kicking off work for all of them.

File: src/Features/Core/Portable/Workspace/BackgroundCompiler.cs [hunk 2]
```diff
@@ -40,141 +54,112 @@ public BackgroundCompiler(Workspace workspace)
...
-
-                    break;
-
-                default:
-                    Rebuild(args.NewSolution, args.ProjectId);
-                    break;
-            }
+            // Stop any work on the current batch and create a token for the next batch.
+            var nextToken = _cancellationSeries.CreateNext();
+            _workQueue.AddWork(nextToken);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Workspace/BackgroundCompiler.cs:88]: Much simpler eventing model.  On any event we always just cancel whatever work we're doing and start the next batch.

File: src/Features/Core/Portable/Workspace/BackgroundCompiler.cs [hunk 3]
```diff
@@ -8,30 +8,43 @@
...
     internal sealed class BackgroundCompiler : IDisposable
     {
-        private Workspace _workspace;
-        private readonly TaskQueue _taskQueue;
+        private Workspace? _workspace;
+        private readonly AsyncBatchingWorkQueue<CancellationToken> _workQueue;
 
         [SuppressMessage("CodeQuality", "IDE0052:Remove unread private members", Justification = "Used to keep a strong reference to the built compilations so they are not GC'd")]
-        private Compilation?[]? _mostRecentCompilations;
+        private readonly ConcurrentSet<Compilation> _mostRecentCompilations = new();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Workspace/BackgroundCompiler.cs:23]: Much easier as a concurrentset.

File: src/Features/Core/Portable/Workspace/BackgroundCompiler.cs [hunk 4]
```diff
@@ -40,141 +53,107 @@ public BackgroundCompiler(Workspace workspace)
...
+            _mostRecentCompilations.Clear();
 
-                _workspace = null!;
+            var workspace = Interlocked.Exchange(ref _workspace, null);
+            if (workspace != null)
+            {
+                workspace.DocumentClosed -= OnDocumentClosed;
+                workspace.DocumentOpened -= OnDocumentOpened;
+                workspace.WorkspaceChanged -= OnWorkspaceChanged;
             }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Workspace/BackgroundCompiler.cs:67]: simpler patterns for doing things once.

File: src/Features/Core/Portable/Workspace/BackgroundCompiler.cs [hunk 5]
```diff
@@ -40,141 +53,107 @@ public BackgroundCompiler(Workspace workspace)
...
-            var allProjectIds = new List<ProjectId>();
-            if (initialProject != null)
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var trackingService = solution.Workspace.Services.GetRequiredService<IDocumentTrackingService>();
+            var visibleProjectIds = trackingService.GetVisibleDocuments().Select(d => d.ProjectId).ToSet();
+            var activeProjectId = trackingService.TryGetActiveDocument()?.ProjectId;
+
+            // Prioritize the project for the active document first.
+            await GetCompilationAsync(activeProjectId).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Workspace/BackgroundCompiler.cs:132]: only processing project one at a time so we don't saturate everything running parallel.

**Review by genlu — APPROVED** [review]

**Pull request merged**